src/extractor/ObeysRuleExtractor.ts

```diff
@@ -0,0 +1,30 @@
+import { ProcessableElementDefinition } from '../processor';
+import { ExportableObeysRule } from '../exportable';
+import { getPath } from '../utils';
+
+export class ObeysRuleExtractor {
+  static process(input: ProcessableElementDefinition): ExportableObeysRule {
+    const invariantKeys: string[] = [];
+    if (input.constraint?.length > 0) {
+      input.constraint.forEach((constraint, i) => {
+        invariantKeys.push(constraint.key);
+        // The ObeysRule only contains the key, but these other attributes
+        // will be handled by the InvariantExtractor. So, they should not be
+        // used for CaretValueRules.
+        input.processedPaths.push(
+          `constraint[${i}].key`,
+          `constraint[${i}].human`,
+          `constraint[${i}].severity`,
+          `constraint[${i}].expression`,
+          `constraint[${i}].xpath`
+        );
```

Comment:
Can't we move this to the `InvariantExtractor` (where it arguably belongs) as long as we ensure that `InvariantExtractor` gets called before the rule extractors?

Reply:
We could, but to do that, we'd need to do some additional refactoring to make rules and invariants get extracted from the same `ProcessableElementDefinition` instances. Right now, `extractRules` and `extractInvariants` create those instances locally, so they disappear along with their `processedPaths` at the end of each function call. Given that the solution in `ObeysRuleExtractor` is admittedly a little magical, what do you think of this idea:
- In the `process` method in `ProfileProcessor` and `ExtensionsProcessor`, build a `ProcessableElementDefinition[]` from `input.differential.element`.
- Modify the `extractRules` and `extractInvariants` methods to take a `ProcessableElementDefinition[]` as a parameter.

That way, these extractors (and any future extractors) can keep the track of `processedPaths` from start to finish. I'll try writing this and see where it gets me.